### PR TITLE
fix(app): Move random mode to navigation menu.

### LINKED ISSUE #537: make random mode more prominent
### Is your feature request related to a problem? Please describe.

I recently found out about the "random" feature, which seems to pop up as a little button on the bottom right. It's pretty neat!

Yet I can't figure out how to bring it back! If I start playing some sounds, it disappears and gets replaced by the "disk" icon to save the presets...

### Describe the solution you'd like

It would be nice if the button would always be there, or be available in the "hamburger menu" for example.

### PR DESCRIPTION
**Pull request by robercoding**
### Changes
Moved random mode to navigation menu!

### Others
- Resolves #537

### CODE REVIEW

**Comment by robercoding** [comment]
I should've added a UI Test.

**Review by ashutoshgngwr — COMMENTED** [review]
Sorry if I wasn't clear earlier, but the intention was to remove the dialog fragment as well and make the random preset fragment like the rest of the screens, e.g. sleep timer.

Wouldn't require changing layouts. Would require to add a new fragment and move existing tests from SoundLibraryFragment.

**Comment by robercoding** [comment]
Oh that makes more sense, I'll change and adapt to that solution which I think it's much better.

**Comment by robercoding** [comment]
Hey @ashutoshgngwr I started doing it now again, and I have some doubts as it seems pretty easy it's also a bit confusing. 

I understand that you want a fragment for random presets that is accesed by navigation menu. Once user click there, what is he expected to see on that fragment screen? 
e.g A blank screen(yet) with a fab button, then click there and a dialog fragment shows up? Like the same random functionality we had on the previous screen but on a new fragment?

PS: I already created a Fragment and user can acceess there but moving test from SoundLibraryFragment to the new RandomPresetFragment test requires more changes to make them work.

Thank you!

**Comment by ashutoshgngwr** [comment]
> I understand that you want a fragment for random presets that is accesed by navigation menu. Once user click there, what is he expected to see on that fragment screen?

Essentially the same layout that was shown in the random preset dialog. When a user clicks the random preset menu item, they are taken to a fragment with the following form.



PS, We can make the play button a floating action button. We can also move and expand the highlighted text to the top like other screens, e.g. _Generated random presets will contain sounds corresponding to the chosen mood. Moreover, dense presets will attempt to play more sounds simultaneously._

> PS: I already created a Fragment and user can acceess there but moving test from SoundLibraryFragment to the new RandomPresetFragment test requires more changes to make them work.

Okay, first let's get everything else right. We will worry about the tests later.

Let me know if I can help you with anything else.

**Comment by robercoding** [comment]
Tell me if you like the changes and what do you want we continue working on? I just added the basics to make sure I'm doing exactly we're looking for.

Let me know how you're looking at them!

File: app/src/main/res/layout/random_preset_fragment.xml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        android:id="@+id/cancel_preset_button"
+        style="?attr/buttonStyle"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="16dp"
+        android:visibility="visible"
+        android:text="@string/cancel"
+        app:layout_constraintEnd_toStartOf="@id/play_preset_button"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintBottom_toBottomOf="parent" />
```
Comment (ashutoshgngwr) [inline app/src/main/res/layout/random_preset_fragment.xml:110]: This isn't really needed...? There's nothing to cancel.

File: app/src/main/res/layout/random_preset_fragment.xml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        android:id="@+id/play_preset_button"
+        style="?attr/buttonStyle"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="16dp"
+        android:visibility="visible"
+        android:text="@string/save"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintBottom_toBottomOf="parent"/>
```
Comment (ashutoshgngwr) [inline app/src/main/res/layout/random_preset_fragment.xml:122]: It can be a floating action button.

File: app/src/main/res/layout/random_preset_fragment.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (ashutoshgngwr) [inline app/src/main/res/layout/random_preset_fragment.xml:2]: would be a good idea to wrap the form in a scroll view. It should be required during landscape orientation.

Also, why didn't you just rename `app/src/main/res/layout/dialog_fragment__random_preset.xml` to `app/src/main/res/layout/random_preset_fragment.xml`?

File: app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      //We'll add later a if player is playing then show else there was an error
+      Toast.makeText(requireContext(), "Playing Random!", Toast.LENGTH_SHORT).show()
+
+      // Your choice to show up in-app review on this fragment!
+      // maybe show in-app review dialog to the user
+      //InAppReviewFlowManager.maybeAskForReview(requireActivity())
+    }
+
+    binding.cancelPresetButton.setOnClickListener {
+      //What do we do cancel? Should we reset preferences button?
```
Comment (ashutoshgngwr) [inline app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt:51]: It's not required in this case.

File: app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      }
+
+      val intensity = when (binding.presetIntensity.checkedRadioButtonId) {
+        R.id.preset_intensity__light -> SoundLibraryFragment.RANGE_INTENSITY_LIGHT
+        R.id.preset_intensity__dense -> SoundLibraryFragment.RANGE_INTENSITY_DENSE
+        else -> SoundLibraryFragment.RANGE_INTENSITY_ANY
+      }
+      MediaPlayerService.playRandomPreset(requireContext(), tag, intensity)
+      //We'll add later a if player is playing then show else there was an error
+      Toast.makeText(requireContext(), "Playing Random!", Toast.LENGTH_SHORT).show()
```
Comment (ashutoshgngwr) [inline app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt:43]: I think clicking the play button should just play a random preset regardless if the player was already playing. The logic remains the same. MediaPlayerService internally stops the existing playback before starting a random (or any) preset. There's no error scenario here per-say.

File: app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+      // Your choice to show up in-app review on this fragment!
+      // maybe show in-app review dialog to the user
+      //InAppReviewFlowManager.maybeAskForReview(requireActivity())
```
Comment (ashutoshgngwr) [inline app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt:47]: It doesn't always result in the dialog being displayed. Sometimes it will be displayed or if the user has already completed the flow once, it won't be shown so it's a safe call.

File: app/src/main/res/layout/random_preset_fragment.xml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+      android:layout_marginTop="20dp"
+      android:layout_marginBottom="7dp"
+      android:text="@string/intensity"
+      android:textAppearance="?attr/textAppearanceHeadline5" />
+
+    <TextView
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_marginVertical="7dp"
+      android:text="@string/intensity_description" />
```
Comment (ashutoshgngwr) [inline app/src/main/res/layout/random_preset_fragment.xml:66]: Move this to the top of the layout. And change the message. The message could be: _Generated random presets will contain sounds corresponding to the chosen mood. Moreover, dense presets will attempt to play more sounds simultaneously._

File: app/src/main/res/layout/random_preset_fragment.xml [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    android:orientation="vertical"
+    android:paddingStart="10dp"
+    android:paddingEnd="10dp">
+
+    <TextView
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_marginBottom="7dp"
+      android:text="@string/mood"
+      android:textAppearance="?attr/textAppearanceHeadline5"
```
Comment (ashutoshgngwr) [inline app/src/main/res/layout/random_preset_fragment.xml:24]: Opinion: This looks ugly in its own fragment. Maybe try headline 6 for both headings.

**Review by ashutoshgngwr — CHANGES_REQUESTED** [review]
I also didn't see any deletions in SoundLibraryFragment. I am guessing you left it out for last.

File: app/src/main/res/layout/random_preset_fragment.xml (hunk 1 above)
Comment (robercoding) [reply app/src/main/res/layout/random_preset_fragment.xml:110]: You're right! I just followed the exact same picture layout, but it was still kinda weird for me looking at the buttons there, it looked pretty ugly, and as you said play button could be a fab! that's why I decided to push these changes before continue doing more work.

File: app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt (hunk 5 above)
Comment (robercoding) [reply app/src/main/java/com/github/ashutoshgngwr/noice/fragment/RandomPresetFragment.kt:43]: True! But I explained myself bad there, let me try again:
Check if is playing like the previous action did its work. If MediaPlayer is not playing (Thing that should never happen after executing `MediaPlayerService.playRandomPreset()`) then show an error. 

Most likely an error will never happen, but just to be sure!

File: app/src/main/res/layout/random_preset_fragment.xml (hunk 3 above)
Comment (robercoding) [reply app/src/main/res/layout/random_preset_fragment.xml:2]: I don't know, didn't think about that simple thing lol, my bad there

**Comment by robercoding** [comment]
Do you like the actual layout? Is the same as before but with implemented changes that you gave me.
Top text _Generated random presets..._ is centered.
Once I finish this I'll continue with sound library fragment deleting its stuff. And moving tests to RandomPresetFragmentTest!

image

Let me know if there's anything I could improve on!

**Comment by ashutoshgngwr** [comment]
@robercoding Looks good. Can be improved I guess. Other fragments have a padding of 20dp on their root layouts (inside scroll view). I think that would look good.

The FAB doesn't look properly positioned on large screens. I am guessing you didn't wrap everything in Relative or Constraint layout, e.g. Relative/Constraint Layout -> FAB + (ScrollView -> Actual Fragment). The bottom-end corner is the conventional position for FABs.

PS, another thing we could try is to orient the radio buttons horizontally. It is easily achievable using FlexboxLayout and we already have it as a [dependency in the timer picker view](https://github.com/ashutoshgngwr/noice/blob/2a0bfb33c8c5cc326dc83c060431547823876650/app/src/main/res/layout/duration_picker_view.xml). The final layout structure would look like RadioGroup -> FlexboxLayout -> RadioButton. Not sure if the results would be good, but definitely worth a shot.

**Comment by robercoding** [comment]
Hey @ashutoshgngwr, I've done these new changes and looks much better now!! 

Using Flexbox scenario with that structure layout (RadioGroup -> Flexbox -> RadioButtons) is impossible because RadioButtons are not controlled by RadioGroup and user can select all 3 options at the same time. But there's an easy way to achieve horizontal orientation by just setting radiogroup orientation as horizontal! I second that idea! ✌️

image

I'll push today as soon as I finish other parts!

**Comment by ashutoshgngwr** [comment]
> Using Flexbox scenario with that structure layout (RadioGroup -> Flexbox -> RadioButtons) is impossible because RadioButtons are not controlled by RadioGroup and user can select all 3 options at the same time. But there's an easy way to achieve horizontal orientation by just setting radio group orientation as horizontal!

Well, using horizontal orientation means that any overflow will be clipped off-screen, so that doesn't work either. :sweat:

On a side note, increase padding between description text and mood heading. It looks a little cluttered right now.

**Comment by robercoding** [comment]
What do you think about using this library for horizontal view?
[Library FlexTools](https://github.com/mrHerintsoaHasina/flextools)

It doesn't overflow if you set more RadioButtons in the future! I couldn't find a way to center its radiobuttons (At least the options I've tried) but setting a minWidth and wrap content in width looks good on both.

Here's a little proof:
image


It's up to you to use this external library! Let me know if you like it.

**Comment by ashutoshgngwr** [comment]
@robercoding It doesn't look mature, but if the library works, I guess it's fine. If something breaks in the future, we will figure something out. :sweat_smile: 

Everything looks great now! :+1:

**Comment by robercoding** [comment]
I've spent time wondering how it works to make play sounds and how it arrives to the UI and to the back until I just remembered you provided us a diagram on CONTRIBUTING.md. My bad there lol.
I've spent time to know if I could add some more UI Tests, but I don't seem to find anything more to add (maybe I'm wrong), let me know if I'm missing something or is that ok for you!

**Review by ashutoshgngwr — APPROVED** [review]
Needs a few styling changes, looks great otherwise. Thanks! :+1:
